File: go/utils/tree.go
package utils

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func NewTree(arr [](int)) *TreeNode {
	if len(arr) == 0 {
		return nil
	}
	idx := 0
	var recurFunc func(node_arr []*TreeNode)
	recurFunc = func(node_arr []*TreeNode) {
		new_node_arr := []*TreeNode{}
		for _, node := range node_arr {
			if node == nil {
				continue
			}
			idx++
			if idx < len(arr) && arr[idx] != -1 {
				node.Left = &TreeNode{Val: arr[idx]}
				new_node_arr = append(new_node_arr, node.Left)
			}
			idx++
			if idx < len(arr) && arr[idx] != -1 {
				node.Right = &TreeNode{Val: arr[idx]}
				new_node_arr = append(new_node_arr, node.Right)
			}
		}
		if len(node_arr) > 0 {
			recurFunc(new_node_arr)
		}
	}
	root := &TreeNode{Val: arr[0]}
	recurFunc([]*TreeNode{root})
	return root
}

func SpreadTree(root *TreeNode) [][]int {
	ans := [][]int{}
	var recurFunc func(root *TreeNode, level int)
	recurFunc = func(root *TreeNode, level int) {
		if len(ans) < level+1 {
			ans = append(ans, []int{})
		}
		if root == nil {
			ans[level] = append(ans[level], -1)
			return
		}
		ans[level] = append(ans[level], root.Val)
		recurFunc(root.Left, level+1)
		recurFunc(root.Right, level+1)
	}
	recurFunc(root, 0)
	fmt.Println("Spreaded Tree:")
	for _, row := range ans {
		fmt.Println(row)
	}
	return ans
}

File: c/utils/tree.c
#include "tree.h"
#include <stdlib.h>
#include <string.h>

void tree_create_recur(int *arr, int lenArr, int idx, tree_node **node_arr,
                       int nodeArrLen) {
  tree_node **new_node_arr = malloc(2 * nodeArrLen * sizeof(tree_node *));
  int m = 0;
  for (int i = 0; i < nodeArrLen; i++) {
    tree_node *node = node_arr[i];
    if (node == NULL) {
      continue;
    }
    idx++;
    if (idx < lenArr && arr[idx != -1]) {
      tree_node *left = malloc(sizeof(tree_node));
      left->val = arr[idx];
      node->left = left;
      new_node_arr[m] = left;
      m++;
    }
    idx++;
    if (idx < lenArr && arr[idx] != -1) {
      tree_node *right = malloc(sizeof(tree_node));
      right->val = arr[idx];
      node->right = right;
      new_node_arr[m] = right;
      m++;
    }
  }
  if (m > 0) {
    tree_create_recur(arr, lenArr, idx, new_node_arr, m);
  }
}

tree_node *tree_create(int *arr, int arrLen) {
  if (arrLen == 0) {
    return NULL;
  }
  tree_node *root = malloc(sizeof(tree_node));
  tree_node *node_arr[1] = {root};
  tree_create_recur(arr, arrLen, 0, &node_arr[0], 1);
  return root;
}
